content/blog/sqlite_parser_pt_2.md

```diff
@@ -0,0 +1,529 @@
+able to determine the total size of the database by multiplying it by
+the `page_size`. The unfortunate truth about this value though is that
+older version of Sqlite do not use this value, which for us means that
+it will often be invalid. Previous version of Sqlite would examine the
+file directly to determine its size so it doesn't invalidate the whole
+file for this to be invalid. There are
+two ways to tell if this value is invalid, first if it is zero otherwise
+we need to look forward to a value called the "version valid for number"
+this will be at byte 92, it should match the value held in `change_counter`
+we've already parsed
+
+The strange thing here is that the documentation _just_ told us that
+the change counter may not be updated if we are in Write Ahead Log 
+mode. This still works because the Write Ahead Log mode and the
+version-valid-for-number were both added in version 3.7.0. That means
+if we were to use an older version of sqlite to modify our database
+it couldn't be in WAL mode, and would always update the change counter.
+At the same time, it will not know about the "version valid for number"
+which will cause these to fall out of sync. 
+
+For us, that means we are going to need to hold off on an additional
+piece of validation for this number until we get almost to the end
+of our header. For right now, we will use an `Option` to represent
+this value, if it was invalid, there is no point in filling the value
+into our struct. We can also use a standard library type `NonZeroU32`
+to enforce the fact that our value cannot be zero. Let's update our
+struct to hold this new value.
+
+```rust
+// header.rs
+use std::num::NonZeroU32;
+
+#[derive(Debug)]
+pub struct DatabaseHeader {
+    pub page_size: PageSize,
+    pub write_version: FormatVersion,
+    pub read_version: FormatVersion,
+    pub change_counter: u32,
+    pub reserved_bytes: u8,
+    /// If populated, the number of pages
+    /// in this database
+    pub database_size: Option<NonZeroU32>,
+} 
+
+```
+
+And again, we can update `parse_header` to capture this value.
+
+```rust
+fn parse_header(bytes; &[u8]) -> Result<DatabaseHeader, Error> {
+    validate_header_string(&contents)?;
+    let page_size = parse_page_size(&contents)?;
+    let write_version = FormatVersion::from(bytes[18]);
+    let read_version = FormatVersion::from(bytes[19]);
+    let reserved_bytes = bytes[20];
+    validate_fraction(bytes[21], 64, "Maximum payload fraction")?;
+    validate_fraction(bytes[21], 32, "Minimum payload fraction")?;
+    validate_fraction(bytes[21], 32, "Leaf fraction")?;
+    let change_counter = crate::try_parse_u32(&bytes[24..28])
+    	.map_err(|msg| Error::InvalidChangeCounter(msg))?;
+   // since parsing a u32 would indicate a much larger error
+   // than this value just being invalid, we will still
+   // return early if this fails
+   let raw_size = crate::try_parse_u32(&bytes[28..32])?;
```

Comment:
I got an error with this as is.
```
error[E0277]: `?` couldn't convert the error to `error::Error`
   --> src/header.rs:137:56
    |
113 | pub fn parse_header(bytes: &[u8]) -> Result<DatabaseHeader, Error> {
    |                                      ----------------------------- expected `error::Error` because of this
...
137 |     let raw_size = crate::try_parse_u32(&bytes[28..32])?;
    |                                                        ^ the trait `std::convert::From<std::string::String>` is not implemented for `error::Error`
    |
    = note: the question mark operation (`?`) implicitly performs a conversion on the error value using the `From` trait
    = note: required by `std::convert::From::from`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0277`.
error: could not compile `sqlite-parser`.

To learn more, run the command again with --verbose.
```

Reply:
So, this one I have actually updated this a bit. Realistically, we wouldn't want to return an error if parsing fails, instead we'd want to just have the value be `None`.